exemptions.py

```diff
@@ -0,0 +1,136 @@
+
+    Parses an open exemptions conf file and returns a two level dict with format:
+
+    {<test_name>:
+        {<test_id>: (<exemption line number>, <exemption expiration>, <exemption reason>)
+        ...
+        }
+    ...
+    }
+
+    The config file consists of:
+
+    comment lines that have the format:
+
+    '#'<anything>\n
+
+    and exemption lines with the format:
+
+    <test name><whitespace><test id><whitespace><exemption expiration><whitespace><exemption reason>\n
+
+    where:
+
+    <anything> := is any non newline character
+    <exemption line number> := line number from the exemption config file
+    <exemption expiration> := date in YYYY-MM-DD format the exemption expires
+    <exemption reason> := reason for the test exemption (e.g. S3 bucket contains public data)
+    <test name> := unparametrized test name
+    <test id> := id from the parametrize decorator
+    <whitespace> := anything str.split i.e. one or more ' ', '\t', or '\n' chars
+
+    Examples:
+
+    >>> from io import StringIO
+    >>> parse_conf_file(StringIO('# comment'))
+    {}
+
+    >>> parse_conf_file(StringIO('test_foo foo-id 2050-01-01 in prod never allow foo\\n'))
+    {'test_foo': {'foo-id': (0, '2050-01-01', 'in prod never allow foo')}}
+
+    >>> parse_conf_file(StringIO(
+    ... 'test_foo foo-id 2050-01-01 in prod never allow foo for foo-id\\n'
+    ... 'test_foo bar-id 2050-01-01 in prod never allow foo for bar-id'
+    ... )) == {
+    ... 'test_foo': {
+    ... 'foo-id': (0, '2050-01-01', 'in prod never allow foo for foo-id'),
+    ... 'bar-id': (1, '2050-01-01', 'in prod never allow foo for bar-id')}}
+    True
+
+
+    Short lines are skipped with a warning:
+
+    >>> parse_conf_file(StringIO('invalid line'))  # doctest:+ELLIPSIS
+    {}
+    >>> # UserWarning: Line 0: Skipping line with fewer than 4 whitespace delimited parts.
+
+    Invalid expirations dates are skipped with a warning:
+
+    >>> parse_conf_file(StringIO('test_foo foo-id 2050-01-AA in prod never allow foo'))
+    {}
+    >>> # UserWarning: Line 0: Skipping line with invalid expiration day '2050-01-AA'
+    >>> parse_conf_file(StringIO('test_foo foo-id 2000-01-01 in prod never allow foo'))
+    {}
+    >>> # UserWarning: Line 0: Skipping line with expiration day in the past '2000-01-01'
+
+    Duplicate test name and IDs are ignored with a warning:
+
+    >>> parse_conf_file(StringIO(
+    ... 'test_foo foo-id 2050-01-01 in prod never allow foo for foo-id\\n'
+    ... 'test_foo foo-id 2051-01-01 in prod never allow foo for bar-id'
+    ... ))
+    {'test_foo': {'foo-id': (0, '2050-01-01', 'in prod never allow foo for foo-id')}}
+    >>> # UserWarning: Line 1: Skipping line with duplicate test name and ID 'test_foo' 'foo-id'
+
+    Does not check that test name and IDs exist (since names might not
+    be collected and IDs can require an HTTP call).
+    """
+    # dict of test name to severity level
+    rules = defaultdict(lambda: {})
```

Comment:
I think you want `defaultdict(dict)`, or am I missing something?

Reply:
They behave the same since it's just looking for a callable, but that's cleaner so I switched to `dict`.

Reply:
oh gotcha, cool cool :)